sdk/cognitiveservices/textanalytics/Azure.AI.TextAnalytics/src/TextAnalyticsServiceSerializer.cs

```diff
@@ -293,12 +301,19 @@ private static RecognizeEntitiesResultCollection ReadRecognizeEntitiesResultColl
                 collection.Add(new RecognizeEntitiesResult(error.Id, error.ErrorMessage));
             }
 
+            collection = SortHeterogeneousCollection(collection, idToIndexMap);
+
             TextBatchStatistics statistics = ReadDocumentBatchStatistics(root);
             string modelVersion = ReadModelVersion(root);
 
             return new RecognizeEntitiesResultCollection(collection, statistics, modelVersion);
         }
 
+        private static List<T> SortHeterogeneousCollection<T>(List<T> collection, Dictionary<string, int> idToIndexMap) where T : TextAnalysisResult
+        {
+            return collection.OrderBy(result => idToIndexMap[result.Id]).ToList();
```

Comment:
`ToArray` is a bit cheaper if you don't actually need a list.

Reply:
Unfortunately, IList is needed for ReadOnlyCollection.